### PR TITLE
test: scheduler supervisor

### PR DESCRIPTION
**Pull request by cutecutecat**
Signed-off-by: cutecutecat <12032261@mail.sustech.edu.cn>

## Description

* 测试结构相关
采用了[此处](https://github.com/golang/mock/issues/352)中的Option 3
*Keeping the mocks in other libs, but writing tests files in "_test" package*
```
/foo
  -- a.go
  -- b.go
  -- some_test.go (package "foo_test")
/testdata
  -- mock_b.go
```

* 关于peer的修改

此处是修复了祖先判断的bug，如果祖先节点是根节点，子孙节点是叶子节点时，`parent, ok := node.GetParent()`此句会提前失效而输出`false`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: pkg/gc/mocks/gc_mock.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (gaius-qi) [inline pkg/gc/mocks/gc_mock.go:1]: Why add gc mock?

File: scheduler/supervisor/cdn_test.go [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+				assert.Nil(err)
+				assert.Equal([]byte("dragonfly-scheduler-test"), mockTask.DirectPiece)
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			patch := tc.mock(t)
+			cdn := NewCDN(mockCDNDynmaicClient, mockPeerManager, mockHostManager)
```
Comment (gaius-qi) [inline scheduler/supervisor/cdn_test.go:497]: `mockCDNDynmaicClient`, `mockPeerManager` and `mockHostManager` used gomock.

File: scheduler/supervisor/test/peer_test.go [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+				assert.Nil(err)
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockHostManager := mock.NewMockHostManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/peer_test.go:473]: gcRaw rename mockGC.

File: scheduler/supervisor/test/peer_test.go [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockHostManager := mock.NewMockHostManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/peer_test.go:519]: gcRaw rename mockGC

File: scheduler/supervisor/test/peer_test.go [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+			gcRaw.EXPECT().Add(gomock.Any()).Return(nil).AnyTimes()
+
+			cfg := config.New()
+			peerManager, _ := supervisor.NewPeerManager(cfg.Scheduler.GC, gcRaw, mockHostManager)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/peer_test.go:581]: must have error check.

File: scheduler/supervisor/test/peer_test.go [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+				assert.True(success)
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockHostManager := mock.NewMockHostManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/peer_test.go:634]: gcRaw rename mockGC

File: scheduler/supervisor/test/task_test.go [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockPeerManager := mock.NewMockPeerManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:380]: gcRaw rename mockGC

File: scheduler/supervisor/test/task_test.go [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+				assert.False(success)
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockHostManager := mock.NewMockPeerManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:432]: gcRaw rename mockGC

File: scheduler/supervisor/test/task_test.go [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+			gcRaw.EXPECT().Add(gomock.Any()).Return(nil).AnyTimes()
+
+			cfg := config.New()
+			taskManager, _ := supervisor.NewTaskManager(cfg.Scheduler.GC, gcRaw, mockHostManager)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:436]: must have error check

File: scheduler/supervisor/test/task_test.go [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockHostManager := mock.NewMockPeerManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
+			gcRaw.EXPECT().Add(gomock.Any()).Return(nil).AnyTimes()
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:482]: gcRaw rename mockGC

File: scheduler/supervisor/test/task_test.go [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+
+			cfg := config.New()
+			taskManager, _ := supervisor.NewTaskManager(cfg.Scheduler.GC, gcRaw, mockHostManager)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:485]: must have error check

File: scheduler/supervisor/test/task_test.go [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+				assert.True(success)
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctl := gomock.NewController(t)
+			defer ctl.Finish()
+			mockPeerManager := mock.NewMockPeerManager(ctl)
+			gcRaw := gc_mock.NewMockGC(ctl)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:537]: gcRaw rename mockGC

File: scheduler/supervisor/test/task_test.go [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+			gcRaw.EXPECT().Add(gomock.Any()).Return(nil).AnyTimes()
+
+			cfg := config.New()
+			taskManager, _ := supervisor.NewTaskManager(cfg.Scheduler.GC, gcRaw, mockPeerManager)
```
Comment (gaius-qi) [inline scheduler/supervisor/test/task_test.go:541]: must have error check

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/supervisor/test/mock/mock_peer.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
+// Code generated by MockGen. DO NOT EDIT.
```
Comment (gaius-qi) [inline scheduler/supervisor/test/mock/mock_peer.go:1]: Rename dir to mocks

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: scheduler/supervisor/cdn_test.go (hunk 2 above)
Comment (cutecutecat) [reply scheduler/supervisor/cdn_test.go]: used in PeerManager and TaskManager initial.

File: scheduler/supervisor/test/task_test.go (hunk 9 above)
Comment (cutecutecat) [reply scheduler/supervisor/test/task_test.go]: inside TaskManager_New

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by 244372610 — APPROVED** [review]

**Pull request merged**